### DISCUSSION TITLE
datetime format

### DISCUSSION THREAD

**Discussion by qn222ad** [highfestiva/finplot#192] [Q&A] [answered]
The strange thing was that if I used data from yahoo finance, the add_line function worked very good. chart plot dates nice and clear. However, yahoo finance did not had all the data I want then I need to use data from csv file.
As I used data from cvs file, then add_line did not work. 
I used the date format as follows: dates= pd.to_datetime(pivot_e.index.date).astype('int64')
What is the potential problem of this? 
Thanks :)

**Comment by highfestiva** [comment, marked as answer]
Don't cast do int64 unless you have to.

**Reply by qn222ad** [reply]
I formatted the time only like this (pd.to_datetime(pivot_e.index.date)) but it does not work. I changed index and datetime to second it still the same. I tried every type of time format I knew but only data rendering online work, not csv :):)

**Reply by highfestiva** [reply]
That's why it's good to supply a complete, minimal example so I can test it. What is your date format before running `to_datetime`?

**Reply by qn222ad** [reply]
data from yahoo finace:
![2021-05-19 23_39_07-Finance Plot](https://user-images.githubusercontent.com/77920707/118888423-ed272600-b8fb-11eb-82e2-8e7d7e9a65a1.png)
data from csv:
![2021-05-19 23_39_49-Finance Plot](https://user-images.githubusercontent.com/77920707/118888453-f57f6100-b8fb-11eb-822b-6ed50f8e20be.png)
In data from csv file, I set both index in data file and csv file of dates using pd.to_datetime(dfn['Date'])
    dfn.loc[:,'Date'] = pd.to_datetime(dfn['Date'], errors='coerce', utc=True)
    dfn = dfn.set_index('Date')
date time show in pivot dates: 
![2021-05-19 23_45_49-astro – small_cycle py](https://user-images.githubusercontent.com/77920707/118888874-8eae7780-b8fc-11eb-82f5-9ec279a98170.png)
and in data: 
![2021-05-19 23_46_20-astro – small_cycle py](https://user-images.githubusercontent.com/77920707/118888908-99690c80-b8fc-11eb-9722-8e0ec9de42fe.png)
I changed both index to second, but still the same... did not know where the problem.
Please advise if you know.
Thanks

**Reply by qn222ad** [reply]
This is example code:
```python
import yfinance as yf
import pandas as pd
pd.set_option('display.max_colwidth', None)
import finplot as fplt

selected = []
def read_days_only(df):
    dfn = df.copy()
    dfn.loc[:,'Date'] = pd.to_datetime(dfn['Date'], errors='coerce')
    dfn = dfn.set_index('Date')
    dfn['Date'] = dfn.index
    return dfn
def extract_price(df1,df2):
    df1 = df1[pd.to_datetime(df1.index.date.astype(str)).isin(pd.to_datetime(df2.index.date))]
    return df1

pivot = read_days_only(pd.read_csv('extracted.csv'))

data = read_days_only(pd.read_csv('data/BankNifty.csv'))
#data from yahoo fianance
#data = pd.DataFrame(yf.download("^NSEBANK",period='max'))

data['Date'] = data.index
data['Date'] = pd.to_datetime(data['Date']).astype('int64')
data['High']=data['High'].values+4 #old data dont have 
data['Low']=data['Low'].values-4 #old data dont have 
mx=2*max(data.High)
pivot_e=extract_price(data,pivot)
dates= pd.to_datetime(pivot_e.index.date)
fplt.candlestick_ochl(data[['Open','Close','High','Low']])
for x, (d, price) in enumerate(zip(dates, pivot_e.Close)):
    fplt.add_line((d, 0), (d, mx), color='#CC9EC1', style='--')
fplt.show()
```

**Reply by highfestiva** [reply]
Send all files in a zip. You're handling the index wrong. You probably want something like this:
```python
data = data.reset_index() # gets you the index into a column
...
data = data.set_index('my_date_column') # set the index back to using datetime again
```
This has nothing to do with finplot, but pandas. It takes some time to learn the basics, but I suggest you do. GL!

**Reply by qn222ad** [reply]
Yes, I am new to coding:D
I checked data type of both 
data = read_days_only(pd.read_csv('data/BankNifty.csv'))
print(data.index.inferred_type == "datetime64")
data = pd.DataFrame(yf.download("^NSEBANK",period='max'))
print(data.index.inferred_type == "datetime64")
 both return True
then they must be at the same type
I will read about datetime in pandas. However, I did not find any different between the two :(

**Reply by highfestiva** [reply]
I ran your example, it shows dates. What is the problem?
![image](https://user-images.githubusercontent.com/160303/118988039-1b9b1480-b981-11eb-8078-ae31da03ccaa.png)

**Comment by qn222ad** [comment]
I can plot them now,
Thank you very much!
I need to change datetime structure (origin dd/mm/yyyy) need to replace / to - then it work

**Reply by qn222ad** [reply]
I think fintplot is quite senditive to data datetime.
I changed other souse of data, use the same datetime format which I did above, then the problem occur again....
the the function I used to plot my old data source ( can do many times) now had error
AssertionError: not possible to plot this primitive without a prior time-range to compare to
What dose this mean? my input was the same as previous, only change different dataset source?
Thanks

**Reply by highfestiva** [reply]
finplot is not _sensitive_ to date, it _requires_ correct timestamps in order to know what to plot. If you print your data first, you can tell if your data contains date+time and if they are correct. Just feeding incorrect data into a library won't work, and shouldn't work.

In this case the library tells you what is wrong: you are trying to plot something without datetime, and you have not supplied datetime in any previous plot either that finplot can use.

**Reply by qn222ad** [reply]
The problem is that the data format is the same too... that why I did not know where did it go wrong. 
![2021-05-25 13_10_39-astro – aspect_extraction py](https://user-images.githubusercontent.com/77920707/119488309-ad789800-bd5a-11eb-8946-c25819423de9.png)
![2021-05-25 13_10_47-astro – aspect_extraction py](https://user-images.githubusercontent.com/77920707/119488312-ad789800-bd5a-11eb-9fe9-1b0c0e1b33ab.png)

**Reply by qn222ad** [reply]
Can I ask what is correct correct timestamps in finplot? datetime64[ns]? 
my data
Open            float64
High            float64
Low             float64
Close           float64
Time     datetime64[ns]
dtype: object
Date     datetime64[ns]
Open            float64
High            float64
Low             float64
Close           float64

**Reply by highfestiva** [reply]
`datetime64[ns]` is fine. You're probably doing something like `plot(df['X'])` without having datetime in the index. See the [tutorial](https://github.com/highfestiva/finplot/wiki/Tutorial-1) in the wiki.

**Reply by qn222ad** [reply]
I always had datetime in index, because i used the index date to link with other data table. You see two pictures above, all was in index. The datetime format take me alot of time, the thing is I do not know where to find error because its format was datetime64[ns] and the same

**Reply by highfestiva** [reply]
The second one has three of each period. That does not constitute a time series. A time series is a bunch of data _in sequence_.

**Reply by qn222ad** [reply]
Thank for good spot, it was error from using pd.merge. However, After fix this, I still had the same issue:
this is my input for line
![2021-05-25 19_41_13-astro – C__Users_Nga Nguyen_Desktop_astro_venv_Lib_site-packages_finplot___init_](https://user-images.githubusercontent.com/77920707/119543927-7a9dc680-bd91-11eb-96c4-1aa388309258.png)
The data is not in sequence:
![2021-05-25 19_41_36-astro – aspect_extraction py](https://user-images.githubusercontent.com/77920707/119543962-87bab580-bd91-11eb-95da-0d83338544ea.png)
and the result:
![2021-05-25 19_44_36-Earth Mars](https://user-images.githubusercontent.com/77920707/119544070-aae56500-bd91-11eb-883a-f896c114b37f.png)
Where did it goes wrong?
print(data.index.inferred_type == "datetime64")
print(pivot.index.inferred_type == "datetime64")

result 
True
True

**Reply by highfestiva** [reply]
So sort it then.
